secret/store.go

```diff
@@ -2,26 +2,81 @@ package secret
 
 import (
 	"context"
+	"sync"
 
 	"github.com/dagger/dagger/core"
 	bkgw "github.com/moby/buildkit/frontend/gateway/client"
 	"github.com/moby/buildkit/session/secrets"
 )
 
 func NewStore() *Store {
-	return &Store{}
+	return &Store{
+		nameToDigest:  map[string]string{},
+		idToPlaintext: map[core.SecretID]string{},
+	}
 }
 
 var _ secrets.SecretStore = &Store{}
 
 type Store struct {
 	gw bkgw.Client
+
+	mu            sync.Mutex
+	nameToDigest  map[string]string
+	idToPlaintext map[core.SecretID]string
 }
 
 func (store *Store) SetGateway(gw bkgw.Client) {
 	store.gw = gw
 }
 
+// AddSecret adds the secret identified by user defined name with its plaintext
+// value to the secret store.
+func (store *Store) AddSecret(_ context.Context, name, plaintext string) (core.SecretID, error) {
+	store.mu.Lock()
+	defer store.mu.Unlock()
+
+	id, err := core.NewSecretID(name, plaintext)
+	if err != nil {
+		return id, err
+	}
+
+	digest, err := id.Digest()
+	if err != nil {
+		// We shouldn't arrive here unless we messed up on our side.
+		// We should stop here to analyze what's wrong.
+		panic(err)
+	}
+
+	// add the digest to the map
+	store.nameToDigest[name] = digest
+
+	// add the plaintext to the map
+	store.idToPlaintext[id] = plaintext
+
+	return id, nil
+}
+
+// GetSecret returns the plaintext from the id.
 func (store *Store) GetSecret(ctx context.Context, id string) ([]byte, error) {
-	return core.NewSecret(core.SecretID(id)).Plaintext(ctx, store.gw)
+	store.mu.Lock()
+	defer store.mu.Unlock()
+
+	secretID := core.SecretID(id)
+
+	// we check if it's the old SecretID format
+	isOldSecretIDFormat, err := secretID.IsOldFormat()
+	if err != nil {
+		return nil, err
+	}
+
+	if isOldSecretIDFormat {
+		// we use the legacy SecretID format
+		return core.NewSecret(core.SecretID(id)).Plaintext(ctx, store.gw)
+	}
+
+	// we use the new SecretID format
+	plaintext := []byte(store.idToPlaintext[secretID])
```

Comment:
Might be better to raise an error of the secret isn't found. I could see that happening if someone copy-pasted a `SecretID` from a previous session (maybe in the playground). An error would be more obvious than silently passing in a blank secret.

Reply:
Very good point. I'm not happy I didn't see it by myself because, clearly those could happen and they would be hard to debug…

Reply:
done in [f705a19](https://github.com/dagger/dagger/pull/4667/commits/f705a195eb5f8ccacc3f62a5bef36b01c3ece780)